src/main/java/apoc/util/Util.java

```diff
@@ -686,4 +691,99 @@ public static boolean classExists(String className) {
+                    tx.success();
+                    return result;
+                }
+                catch ( Exception e )
+                {
+                    throw e;
+                }
+            } );
+            Thread threadForTransaction = new Thread( resultFuture );
+
+            try
+            {
+                threadForTransaction.start();
+                threadForTransaction.join();
+
+                // Query has not finished yet, wait.
+                while ( !resultFuture.isDone() )
+                {
+                    LockSupport.parkNanos( 1000 );
+                }
+
+                return resultFuture.get().stream().map( MapResult::new );
+            }
+            catch ( Exception e )
+            {
+                Throwable cause = e.getCause();
+
+                if ( cause instanceof QueryExecutionException )
+                {
+
+                    String statusCode = ((QueryExecutionException) cause).getStatusCode();
+                    if ( !isRetriableException( statusCode, additionalRetryCodes ) )
+                    {
+                        transactionFailureMessage = "Failed after " + timesTried + " out of " + retries + " retries. QueryExecutionException : " +
+                                ((QueryExecutionException) cause).getStatusCode();
+                        break;
+                    }
+                }
+
+                log.warn( "Retrying operation " + timesTried + " of " + retries );
+                Util.sleep( 100 );
+                timesTried++;
+            }
+            transactionFailureMessage = "Failed after " + timesTried + " of " + retries + " retries.";
+        }
+        while ( timesTried <= retries );
+
+        throw new TransactionFailureException( transactionFailureMessage );
+    }
+
+    private interface RetriableExceptions extends Status
+    {
+        List<Status> QueryExecutionExceptionStatusCodeList = Arrays.asList(
+                General.UnknownError,
+                Network.CommunicationError,
+                Procedure.ProcedureTimedOut,
+                Schema.ConstraintValidationFailed,
+                Schema.SchemaModifiedConcurrently,
+                Statement.ConstraintVerificationFailed,
+                Transaction.ConstraintsChanged,
+                Transaction.DeadlockDetected,
+                Transaction.InstanceStateChanged,
+                Transaction.LockAcquisitionTimeout,
+                Transaction.LockSessionExpired,
+                Transaction.Outdated,
+                Transaction.TransactionCommitFailed,
+                Transaction.TransactionStartFailed,
+                Transaction.TransactionTimedOut
+        );
+    }
+
+    private static boolean isRetriableException( String statusCode, List<String> customRetriable )
+    {
+
+        if ( RetriableExceptions.QueryExecutionExceptionStatusCodeList.stream().map( sc -> sc.code().serialize().equals( statusCode ) ).reduce( ( x, acc ) -> acc || x ).get()
```

Comment:
probably a loop + if with early return is easier here ?
there is also any/none() etc on streams

Reply:
I'll fix this up.